### PR TITLE
[LTC] Add custom lazy tensor save function

### PR DESCRIPTION
**Pull request by antoniojkim**
We need a custom `save` function for checkpointing a lazy model, similar to what exists in PyTorch/XLA:
https://github.com/pytorch/xla/blob/3eb8a9d9eb4ebb0b064461c3704650241625654e/torch_xla/core/xla_model.py#L994
The purpose of this function is to move any lazy tensors to CPU before saving the checkpoint.

The way I implemented it was to create a general structure visitor, adapted from a function that we use quite often in Cerebras internal repositories. If there is a better tool already available in PyTorch that does the same things, I'm open to suggestions.

CC: @wconstab @Krovatkin @JackCaoG

### CODE REVIEW

**Comment by alanwaketan** [comment]
Cc myself and @JackCaoG.

**Comment by antoniojkim** [comment]
@wconstab @JackCaoG Gentle reminder to please review when you get a chance

File: torch/_lazy/visit.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
+import torch
+
+
+def visit_structure(structure, select_fn, apply_fn, scope=None, strict=True):
+    """Recursively traverse nested structureure and calls the apply_fn callable on
```
Comment (wconstab) [inline torch/_lazy/visit.py:5]: hmm there is this 'pytree flatten' infra that is available. i am wondering if it is directly usable for you or not.

https://github.com/pytorch/pytorch/blob/b8b54eccd293032ed6761ac21c72b1adf27999d3/torch/utils/_pytree.py

I think if you just want to hit all the tensors with your 'apply_fn' you can directly use pytree and not have to write all this code.

Comment (antoniojkim) [reply torch/_lazy/visit.py]: Thanks for the suggestion. I wasn't aware of this util

File: torch/_lazy/__init__.py [hunk 2]
```diff
@@ -34,3 +35,15 @@ def sync_multi(tensors, devices):
 def get_tensor_id(tensor):
     """Return a unique id of the lazy tensor maintained by LTC"""
     return torch._C._lazy._get_tensor_id(tensor)
+
+
+def to_cpu(tensors, devices=None):
+    devices = devices or ["lazy"]
+
+    flattened, spec = tree_flatten(tensors)
```
Comment (desertfire) [inline torch/_lazy/__init__.py:43]: `tree_map` would be a good fit for your purpose here. Example code, https://github.com/pytorch/pytorch/blob/abcf01196cd27805349aa892db847f9a61f52c0e/functorch/functorch/_src/aot_autograd.py#L212

Comment (antoniojkim) [reply torch/_lazy/__init__.py:43]: I specifically chose to manually flatten and unflatten as I need the flattened version to call `sync_multi` and it seemed unnecessary to call "flatten" again indirectly via `tree_map`

**Comment by antoniojkim** [comment]
@wconstab can I get another review for this?

**Review by wconstab — APPROVED** [review]
thanks for refactoring to use pytree, lgtm.

**Pull request closed without merging**